pkg/slayers/path/scion/raw_spec.gobra

```diff
@@ -192,12 +207,26 @@ requires 0 <= idx && idx <= unfolding acc(r.Mem(ubuf), p) in len(r.Raw)
 ensures  acc(&r.Raw, p/2)
 ensures  r.Raw === old(unfolding acc(r.Mem(ubuf), p) in r.Raw)
 ensures  acc(slices.AbsSlice_Bytes(r.Raw[:idx], 0, idx), p) && acc(r.Base.Mem(), p/2)
+ensures  forall i int :: 0 <= i && i < idx ==> (unfolding acc(slices.AbsSlice_Bytes(r.Raw[:idx], 0, idx), p) in &r.Raw[:idx][i] == &r.Raw[i] && r.Raw[i] == old(unfolding acc(r.Mem(ubuf), p) in unfolding acc(slices.AbsSlice_Bytes(ubuf, 0, len(ubuf)), p) in r.Raw[i]))
+ensures  (unfolding acc(r.Base.Mem(), p/2) in r.Base.PathMeta.CurrINF) == old(unfolding acc(r.Mem(ubuf), p) in unfolding acc(r.Base.Mem(), p) in r.Base.PathMeta.CurrINF)
+ensures  (unfolding acc(r.Base.Mem(), p/2) in r.Base.PathMeta.SegLen[0]) == old(unfolding acc(r.Mem(ubuf), p) in unfolding acc(r.Base.Mem(), p) in r.Base.PathMeta.SegLen[0])
+ensures  (unfolding acc(r.Base.Mem(), p/2) in r.Base.PathMeta.SegLen[1]) == old(unfolding acc(r.Mem(ubuf), p) in unfolding acc(r.Base.Mem(), p) in r.Base.PathMeta.SegLen[1])
+ensures  (unfolding acc(r.Base.Mem(), p/2) in r.Base.PathMeta.SegLen[2]) == old(unfolding acc(r.Mem(ubuf), p) in unfolding acc(r.Base.Mem(), p) in r.Base.PathMeta.SegLen[2])
+ensures  (unfolding acc(r.Base.Mem(), p/2) in r.Base.NumINF) == old(unfolding acc(r.Mem(ubuf), p) in unfolding acc(r.Base.Mem(), p) in r.Base.NumINF)
 ensures  r.RawIdxPermRemainder(ubuf, idx, p)
 decreases
 func (r *Raw) RawIdxPerm(ubuf []byte, idx int, p perm) {
 	unfold acc(r.Mem(ubuf), p)
-	slices.SplitByIndex_Bytes(ubuf, 0, len(ubuf), idx, p)
-	slices.Reslice_Bytes(ubuf, 0, idx, p)
+	unfold acc(slices.AbsSlice_Bytes(ubuf, 0, len(ubuf)), p/2)
+	slices.SplitByIndex_Bytes(ubuf, 0, len(ubuf), idx, p/2)
+	unfold acc(slices.AbsSlice_Bytes(ubuf, 0, idx), p/2)
```

Comment:
we should add a comment here explaining why we do this in two steps

Reply:
On a second thought, we go out of our way to ensure framing in a performant way (e.g., by not introducing any quantified assertions). I wonder if, instead of doing this in steps, which is rather ugly and unmaintainable, we could use `AbsSliceC_Bytes` and its corresponding methods internally (i.e., without exposing this in the interfaces of the public method's specs). In particular, if we start in AbsC(s, ..., contents) and end up with the same predicate instance, we know that the contents of the slice did not change.

Reply:
additionally, we could have conversion functions between the two kinds of predicates